src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -229,54 +230,76 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-    } andThen {
-      DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceMethodConfigs(workspaceContext.workspaceId))
-    } andThen {
-      dataAccess.workspaceQuery.deleteWorkspaceEntitiesAndAttributes(workspaceContext.workspaceId)
-    } andThen {
-      DBIO.seq(workspaceContext.workspace.accessLevels.map { case (_, group) =>
-        dataAccess.rawlsGroupQuery.delete(group)
-      }.toSeq: _*)
-    } andThen {
-      dataAccess.workspaceQuery.delete(workspaceName)
-    } andThen {
-      DBIO.successful(RequestComplete(StatusCodes.Accepted, s"Your Google bucket ${workspaceContext.workspace.bucketName} will be deleted within 24h."))
+    val x: Future[(Seq[WorkflowRecord], String, Seq[Option[RawlsGroup]])] = dataSource.inTransaction { dataAccess => {
+      for {
+        workflowsToAbort <- dataAccess.workflowQuery.findActiveWorkflowsWithExternalIds(workspaceContext)
+        _ <- dataAccess.workflowQuery.findWorkflowsByWorkspace(workspaceContext).result.map { recs => recs.collect {
+          case wf if !WorkflowStatuses.withName(wf.status).isDone && wf.externalId.isDefined => executionServiceCluster.abort(wf, userInfo).map {
+            case Failure(regrets) =>
+              logger.info(s"failure aborting workflow ${wf.externalId} while deleting workspace ${workspaceName}", regrets)
+              Failure(regrets)
+            case success => success
+          }
+          //If a workflow is not done, automatically changed its status to Aborted
+          case wf if !WorkflowStatuses.withName(wf.status).isDone => dataAccess.workflowQuery.updateStatus(wf, WorkflowStatuses.Aborted)
+        }
+        }
+        // Instead of deleting bucket, send message to delete bucket to BucketDeletionMonitor
+        _ <- DBIO.successful(bucketDeletionMonitor ! BucketDeletionMonitor.DeleteBucket(workspaceContext.workspace.bucketName))
+
+        groupRefs: Set[RawlsGroupRef] = workspaceContext.workspace.accessLevels.values.toSet ++ workspaceContext.workspace.realmACLs.values
+        groupsToRemove <- DBIO.sequence(groupRefs.toSeq.map (groupRef => dataAccess.rawlsGroupQuery.load(groupRef)))
+
+        _ <- DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceAccessReferences(workspaceContext.workspaceId))
+        _ <- DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceSubmissions(workspaceContext.workspaceId))
+        _ <- DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceMethodConfigs(workspaceContext.workspaceId))
+        _ <- dataAccess.workspaceQuery.deleteWorkspaceEntitiesAndAttributes(workspaceContext.workspaceId)
+        _ <- {
+          DBIO.seq(workspaceContext.workspace.accessLevels.map { case (_, group) =>
+            dataAccess.rawlsGroupQuery.delete(group)
+          }.toSeq: _*)
+        }
+        _ <- dataAccess.workspaceQuery.delete(workspaceName)
+      } yield {
+        (workflowsToAbort, workspaceContext.workspace.bucketName, groupsToRemove)
+      }
+    }
+    }
+    for {
```

Comment:
ok, so the above `inTransaction` is the locally-executed, transaction-y block, and this block is the "go off and talk to external services" block. i think i'd function-ise these? i'm mentally comparing this code to `WorkflowSubmissionActor.submitWorkflowBatch` as a "big ole for-comp". but on further inspection i used some, but not zillions, of helper functions just to make the code more readable.

maybe not. consider this just me rambling.